README.md
<pre>
			      dP                     dP   dP                  dP                     
			      88                     88   88                  88                     
			.d888b88 .d8888b. 88d888b. d8888P 88d888b. .d8888b. d8888P .d8888b. 88d888b. 
			88'  `88 88'  `88 88'  `88   88   88'  `88 Y8ooooo.   88   88ooood8 88'  `88 
			88.  .88 88.  .88 88         88   88    88       88   88   88.  ... 88.  .88 
			`88888P8 `88888P8 dP         dP   dP    dP `88888P'   dP   `88888P' 88Y888P' 
			                                                                    88       
			                                                                    dP       
</pre>

An arduino Sampler/Synthesizer.

This started (an continues) as a learning process... but it turned out to be a pretty damn dubstep-ish music instrument. xPP

**Synth Features:**

- Two independent 3D synths. (possible even more)
- Multiple simultaneous waves per synth. (Square, Triangle, Saw or Sine)
- Multiple pre-programed scales (intervals).
- Octave/Scale selection.
- Pitch/Gain/ChainSaw alteration with touch sensor, accelerometer and/or photoresistor. (6 axes)
- Midi Out.

**Sampler Features:**

- 9 Samples. 8Bits. 8khz.
- Editable 2 bars beats. 64 steps per sample (32 each bar).
- Variable BPM.
- Multiple quantization presets.
- Midi Out.

It also has a 3 channel mixer UI with a gain slider and a mute button per channel.

Everything runs all together on an arduino mega with a 9V battery and is controlled via a graphical touch interface (in a 320x240 TFT) and a few inputs.

Both the synth and the sampler allow to record/save/load sequences into an SD Card.

The midi outs are on separate midi channels for convenience, but (like ohter config options) the channel numbers are constants at the top of the arduino sketch.

Inspiration (and some code & schematics) for this project came from:

- http://code.google.com/p/tinkerit/wiki/Auduino
- http://www.instructables.com/id/Electronic-Instrument

**License:**

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the [GNU General Public License](https://github.com/danielesteban/DarthStep/blob/master/LICENSE)
along with this program. If not, see <http://www.gnu.org/licenses/>.


CHANGELOG.md
## v2.5 (Feb 06 2013)

  - Sequencer UIView
  - SequencableUI Interface
  - Save/load/clear unified for sampler and synths with new file browser.
  - SD final implementation with hotswap.
  - Memory optimizations. also added more debug capabilities.
  - Bugfixes & improvements. Fixed some memory leaks.

## v2.4 (Feb 01, 2013)
  
  - PhotoResistor support as another synth axis.
  - SetNote refactoring. (removed old selectedNote/note behaviour)
  - #ifdefs refactoring.
  - Fritzing files updated.

## v2.3 (Jan 29, 2013)
  
  - 3D Synths! (I'd say 5D ;P)
  - Accelerometer hardware support.
  - Beta version of autoOrientation.
  - Sampler toggle/clear.
  - Added Makefile.
  - Some bugfixes and improvements.

## v2.2 (Jan 28, 2013)

  - Mixer UI with sliders & mute buttons for each channel.
  - Cool atari like intro UIView.
  - License UIView.
  - Some bugfixes and improvements.
  - Added CHANGELOG.

## v2.1 (Jan 26, 2013)

  - SynthConfig Dialog really improved.. also added the "truly 2D" :PP which means now you can decide which axis modifies which synth property.
  - MidiOut added to both sampler & synth.
  - Optimized Wave library & Samples.
  - A bunch of bugfixes and improvements.

## v2.0 (Jan 23, 2013)

  - The very first of the TFT Touch versions... Already had:
    - The sampler
    - The synths.
    - The sequencer.
  - I Also added the GNU GPL.

## v1.0 (Dec 29, 2012)

  - This were the first commits into github.. the old GLCD version.


License.cpp
/*
    License.cpp - DarthStep license UIView.
    Created by Daniel Esteban, January 26, 2013.
*/

#include "License.h"

prog_char licenseText1[] PROGMEM = "DarthStep | (C) 2013 Daniel Esteban N.";
prog_char licenseText2[] PROGMEM = "This program is free software: you can";
prog_char licenseText3[] PROGMEM = "redistribute it and/or modify it under";
prog_char licenseText4[] PROGMEM = "the terms of the GNU General Public";
prog_char licenseText5[] PROGMEM = "License as published by the Free";
prog_char licenseText6[] PROGMEM = "Software Foundation, either version 3";
prog_char licenseText7[] PROGMEM = "of the License, or (at your option)";
prog_char licenseText8[] PROGMEM = "any later version.";
prog_char licenseText9[] PROGMEM = "This program is distributed in the";
prog_char licenseText10[] PROGMEM = "hope that it will be useful, but";
prog_char licenseText11[] PROGMEM = "WITHOUT ANY WARRANTY; without even the";
prog_char licenseText12[] PROGMEM = "implied warranty of MERCHANTABILITY or";
prog_char licenseText13[] PROGMEM = "FITNESS FOR A PARTICULAR PURPOSE. See";
prog_char licenseText14[] PROGMEM = "the GNU General Public License for";
prog_char licenseText15[] PROGMEM = "more details.";

PROGMEM const char * licenseTexts[] = {
	licenseText1,
	licenseText2,
	licenseText3,
	licenseText4,
	licenseText5,
	licenseText6,
	licenseText7,
	licenseText8,
	licenseText9,
	licenseText10,
	licenseText11,
	licenseText12,
	licenseText13,
	licenseText14,
	licenseText15
};

License::License() {
	availableOrientations[PORTRAIT] = 0;
}

void License::render(UTFT tft) {
	tft.clrScr();
	tft.setFont(SmallFont);
	_tft = tft;
	_index = _offset = 0;
	update();
}

void License::update() {
	if(_index == 15) return;

	char buffer[64];
	strcpy_P(buffer, (char *) pgm_read_word(&(licenseTexts[_index])));
	if(_index == 0) {
		_tft.setColor(255, 255, 255);
		_tft.fillRect(0, 0, _tft.getDisplayXSize() - 1, 12);
		_tft.setBackColor(255, 255, 255);
		_tft.setColor(0, 0, 0);
		_tft.print(buffer, 10, 0);

		_tft.setBackColor(0, 0, 0);
		_tft.setColor(255, 255, 255);
	} else {
		_index == 8 && (_offset += 15);
		_tft.print(buffer, 10, (_index * 15) + _offset);
	}
	_index++;
}
